pkl-core/src/main/java/org/pkl/core/ast/builder/AstBuilderNew.java

```diff
@@ -0,0 +1,3043 @@
+    var scope = symbolTable.getCurrentScope();
+    var levelsUp = 0;
+
+    while (parent instanceof If
+        || parent instanceof Trace
+        || parent instanceof Let letExpr && letExpr.getExpr() == child) {
+
+      if (parent instanceof Let) {
+        assert scope != null;
+        scope = scope.getParent();
+        levelsUp += 1;
+      }
+      child = parent;
+      parent = parent.parent();
+    }
+
+    assert scope != null;
+
+    if (parent instanceof ClassPropertyEntry
+        || parent instanceof ObjectProperty
+        || parent instanceof ObjectBodyProperty) {
+      inferredParentNode =
+          InferParentWithinPropertyNodeGen.create(
+              createSourceSection(expr.newSpan()),
+              scope.getName(),
+              levelsUp == 0 ? new GetOwnerNode() : new GetEnclosingOwnerNode(levelsUp));
+    } else if (parent instanceof ObjectElement
+        || parent instanceof ObjectEntry objectEntry && objectEntry.getValue() == child) {
+      inferredParentNode =
+          ApplyVmFunction1NodeGen.create(
+              ReadPropertyNodeGen.create(
+                  createSourceSection(expr.newSpan()),
+                  Identifier.DEFAULT,
+                  levelsUp == 0 ? new GetReceiverNode() : new GetEnclosingReceiverNode(levelsUp)),
+              new GetMemberKeyNode());
+    } else if (parent instanceof ClassMethod || parent instanceof ObjectMethod) {
+      var isObjectMethod =
+          parent instanceof ObjectMethod
+              || parent.parent() instanceof ModuleContext && moduleInfo.isAmend();
+      Identifier scopeName = scope.getName();
+      inferredParentNode =
+          isObjectMethod
+              ? new InferParentWithinObjectMethodNode(
+                  createSourceSection(expr.newSpan()),
+                  language,
+                  scopeName,
+                  levelsUp == 0 ? new GetOwnerNode() : new GetEnclosingOwnerNode(levelsUp))
+              : new InferParentWithinMethodNode(
+                  createSourceSection(expr.newSpan()),
+                  language,
+                  scopeName,
+                  levelsUp == 0 ? new GetOwnerNode() : new GetEnclosingOwnerNode(levelsUp));
+    } else if (parent instanceof Let letExpr && letExpr.getBindingExpr() == child) {
+      // TODO (unclear how to infer type now that let-expression is implemented as lambda
+      // invocation)
+      throw exceptionBuilder()
+          .evalError("cannotInferParent")
+          .withSourceSection(createSourceSection(expr.newSpan()))
+          .build();
+    } else {
+      throw exceptionBuilder()
+          .evalError("cannotInferParent")
+          .withSourceSection(createSourceSection(expr.newSpan()))
+          .build();
+    }
+
+    return doVisitObjectBody(expr.getBody(), inferredParentNode);
+  }
+
+  @Override
+  public ExpressionNode visitAmendsExpr(Amends expr) {
+    // parentExpr is always New, Amends or Parenthesized. The parser makes sure of it
```

Comment:
I like this comment, but it would be useful to link through to the method where this behaviour is expected in the parser.